File: apps/dotnet/FizzBuzz/Domain/Model/FizzBuzzValue.cs
namespace FizzBuzz.Domain.Model;

public sealed class FizzBuzzValue : IEquatable<FizzBuzzValue>
{
    public int Number { get; }
    public string Value { get; }

    public FizzBuzzValue(int number, string value)
    {
        Number = number;
        Value = value;
    }

    public override string ToString() => $"{Number}:{Value}";

    public bool Equals(FizzBuzzValue? other)
    {
        if (other is null) return false;
        if (ReferenceEquals(this, other)) return true;
        return Number == other.Number && Value == other.Value;
    }

    public override bool Equals(object? obj) => Equals(obj as FizzBuzzValue);

    public override int GetHashCode() => HashCode.Combine(Number, Value);
}

File: apps/typescript/src/fizzbuzz/domain/model/fizz-buzz-value.ts
export class FizzBuzzValue {
  private readonly _value: string;
  private readonly _number: number;

  constructor(value: string, number: number) {
    this._value = value;
    this._number = number;
  }

  get value(): string {
    return this._value;
  }

  get number(): number {
    return this._number;
  }

  equals(other: FizzBuzzValue): boolean {
    if (!(other instanceof FizzBuzzValue)) {
      return false;
    }
    return this._value === other._value && this._number === other._number;
  }

  toString(): string {
    return this._value;
  }
}
